File: lintcode/Min_Stack.cpp
class MinStack {
private:
    stack<int> dataStack;
    stack<int> minStack;   
public:
    MinStack() {}

    void push(int number) {
        dataStack.push(number);
        if(minStack.empty() || number <= minStack.top()){
            minStack.push(number);
        }
    }


    int pop() {
        int x = dataStack.top();
        dataStack.pop();
        
        if(x == minStack.top()){
            minStack.pop();
        }
        return x;
    }


    int min() {
        return minStack.top();
    }
};

File: lintcode/java/MinStack.java
public class MinStack {
    private final Deque<Integer> minStack = new ArrayDeque<>();
    private final Deque<Integer> dataStack = new ArrayDeque<>();
    
    public MinStack() {}

    public void push(int x) {
        dataStack.push(x);
        
        if(minStack.isEmpty() || x <= minStack.peek()) minStack.push(x);
    }

    public int pop() {
        if(minStack.isEmpty() || dataStack.isEmpty()) {
            throw new IllegalArgumentException("xxxx");
        }
        int x = dataStack.pop();
        
        if (x == minStack.peek()) minStack.pop();
        return x;
    }

    public int min() {
        if(minStack.isEmpty()) throw new IllegalArgumentException("xxxx");
        return minStack.peek();
    }
}

